gpflow/models/training_mixins.py

```diff
@@ -0,0 +1,58 @@
+import abc
+import collections
+from typing import Callable, Optional, Tuple, TypeVar, Union
+
+
+import tensorflow as tf
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData, OutputData)
+
+
+class InternalDataTrainingLossMixin:
+    def training_loss(self) -> tf.Tensor:
+        """
+        Training loss for models that encapsulate their data.
+        """
+        return -self.maximum_a_posteriori_objective()
```

Comment:
I spoke with @condnsdmatters about making this an abstract class with an abstract `maximum_a_posteriori_objective` method. I think it's the right thing to do, eric seemed far from convinced. I don't want it to block this PR cos it doesn't change any functionality and can very easily be added later.

In the mean time though, I would like the docs to explain that an `maximum_a_posteriori_objective` with a particular signature must be implemented for classes that mix in this, err..  mixin

Reply:
A mixin by definition doesn't make sense without a class to mix it into. All GPflow models subclass `BayesianModel`, which provides the maximum_a_posteriori_objective (note that the abstract method to implement is maximum_likelihood_objective). Adding an abstract method in the mixin would mean people had to be very careful about the multiple-inheritance order, so I am strongly against adding that (now or in the future).

Reply:
I've added a module-level docstring to clarify. Please re-open if you think that's still not clear enough.